File: src/collections_and_containers/cpp/heap/src/heap.cpp
#include "heap.h"
#include <stdexcept>

template <class T>
Heap<T>::Heap(std::function<bool(T, T)> comp) : comp(comp) {}

template <class T>
Heap<T>::Heap(const Heap &other) : data(other.data), comp(other.comp) {}

template <class T> Heap<T> &Heap<T>::operator=(const Heap &other) {
  if (this != &other) {
    data = other.data;
    comp = other.comp;
  }
  return *this;
}

template <class T> void Heap<T>::push(const T &val) {
  data.push_back(val);
  int index = static_cast<int>(data.size()) - 1;
  int parent = parentIdx(index);

  // Bubble up
  while (index > 0 && comp(data[index], data[parent])) {
    std::swap(data[index], data[parent]);
    index = parent;
    parent = parentIdx(index);
  }
}

template <class T> T Heap<T>::pop() {
  if (data.empty()) {
    throw std::out_of_range("Heap is empty");
  }

  T val = data[0];
  data[0] = data.back();
  data.pop_back();

  if (!data.empty()) {
    // Bubble down
    int index = 0;
    int size = static_cast<int>(data.size());

    while (true) {
      int left = leftNodeIdx(index);
      int right = rightNodeIdx(index);
      int smallest = index;

      if (left < size && comp(data[left], data[smallest])) {
        smallest = left;
      }
      if (right < size && comp(data[right], data[smallest])) {
        smallest = right;
      }

      if (smallest == index) {
        break;
      }

      std::swap(data[index], data[smallest]);
      index = smallest;
    }
  }

  return val;
}

template <class T> const T &Heap<T>::peek() const {
  if (data.empty()) {
    throw std::out_of_range("Heap is empty");
  }
  return data[0];
}

template <class T> unsigned int Heap<T>::size() const {
  return static_cast<unsigned int>(data.size());
}

template <class T> bool Heap<T>::empty() const { return data.empty(); }

template <class T> void Heap<T>::clear() { data.clear(); }

template <class T> bool Heap<T>::isMaxHeap() const { return comp(2, 1); }

template <class T> bool Heap<T>::isMinHeap() const { return comp(1, 2); }

template <class T> int Heap<T>::parentIdx(int index) const {
  return (index - 1) / 2;
}

template <class T> int Heap<T>::leftNodeIdx(int index) const {
  return 2 * index + 1;
}

template <class T> int Heap<T>::rightNodeIdx(int index) const {
  return 2 * index + 2;
}

template <class T> MinHeap<T>::MinHeap() : Heap<T>(std::less<T>()) {}

template <class T> MaxHeap<T>::MaxHeap() : Heap<T>(std::greater<T>()) {}

template class Heap<int>;
template class Heap<float>;
template class Heap<double>;
template class Heap<char>;

template class MinHeap<int>;
template class MinHeap<float>;
template class MinHeap<double>;
template class MinHeap<char>;

template class MaxHeap<int>;
template class MaxHeap<float>;
template class MaxHeap<double>;
template class MaxHeap<char>;

File: src/collections_and_containers/python/heap/src/heap.py
"""Heap data structure implementations (min-heap and max-heap)."""

from __future__ import annotations

import abc
from typing import Any, Callable


class Heap(abc.ABC):
    """Abstract base class for heap data structures."""

    def __init__(self, comp: Callable[[Any, Any], bool]) -> None:
        self.data: list[Any] = []
        self.comp = comp

    def __copy__(self) -> Heap:
        new_heap = self.__class__.__new__(self.__class__)
        new_heap.comp = self.comp
        new_heap.data = self.data[:]
        return new_heap

    def push(self, val: Any) -> None:
        """Add an element to the heap."""
        self.data.append(val)
        self._bubble_up(self.size() - 1)

    def pop(self) -> Any:
        """Remove and return the top element."""
        if self.empty():
            raise IndexError("Cannot pop from an empty heap")
        self._swap(0, self.size() - 1)
        val = self.data.pop()
        self._bubble_down(0)
        return val

    def peek(self) -> Any:
        """Return the top element without removing it."""
        if self.empty():
            raise IndexError("Cannot peek at an empty heap")
        return self.data[0]

    def size(self) -> int:
        """Return the number of elements in the heap."""
        return len(self.data)

    def empty(self) -> bool:
        """Check if the heap is empty."""
        return self.size() == 0

    def is_max_heap(self) -> bool:
        """Check if this is a max-heap."""
        return self.comp(2, 1)

    def is_min_heap(self) -> bool:
        """Check if this is a min-heap."""
        return self.comp(1, 2)

    def _bubble_up(self, index: int) -> None:
        if index == 0:
            return
        parent = self._parent(index)
        if self.comp(self.data[index], self.data[parent]):
            self._swap(index, parent)
            self._bubble_up(parent)

    def _bubble_down(self, index: int) -> None:
        left = self._left(index)
        right = self._right(index)
        smallest = index
        if left < self.size() and self.comp(self.data[left], self.data[smallest]):
            smallest = left
        if right < self.size() and self.comp(self.data[right], self.data[smallest]):
            smallest = right
        if smallest != index:
            self._swap(index, smallest)
            self._bubble_down(smallest)

    def _parent(self, index: int) -> int:
        return (index - 1) // 2

    def _left(self, index: int) -> int:
        return 2 * index + 1

    def _right(self, index: int) -> int:
        return 2 * index + 2

    def _swap(self, i: int, j: int) -> None:
        self.data[i], self.data[j] = self.data[j], self.data[i]

    def __eq__(self, other: object) -> bool:
        if isinstance(other, self.__class__):
            return self.data == other.data
        return False

    def __str__(self) -> str:
        return str(self.data)


class MinHeap(Heap):
    """A min-heap where the smallest element is at the top."""

    def __init__(self) -> None:
        super().__init__(comp=lambda x, y: x < y)


class MaxHeap(Heap):
    """A max-heap where the largest element is at the top."""

    def __init__(self) -> None:
        super().__init__(comp=lambda x, y: x > y)
